pkg/reconciler/namespacecontroller/namespace_controller.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package namespacecontroller
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"kpt.dev/configsync/pkg/status"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// NamespaceController is a controller that watches for Namespace events and
+// sets a flag to indicate whether the Reconciler thread needs to reconcile for
+// this namespace event.
+type NamespaceController struct {
+	client client.Client
+	state  *State
+}
+
+// New instantiates the namespace controller.
+func New(cl client.Client, state *State) *NamespaceController {
+	return &NamespaceController{
+		client: cl,
+		state:  state,
+	}
+}
+
+// Reconcile processes the Namespace request to set the flag to indicate whether
+// a new reconciliation needs to be executed by the Reconciler thread.
+func (nc *NamespaceController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ns := &corev1.Namespace{}
+	if err := nc.client.Get(ctx, req.NamespacedName, ns); err != nil {
+		if apierrors.IsNotFound(err) {
+			if nc.state.isSelectedNamespace(req.Name) {
+				// A previously selected namespace is deleted, so set the namespace event.
+				nc.state.setSyncPending()
```

Comment:
Is the cache (`selectedNamespaces`) updated to remove this namespace?

Reply:
Nope. `selectedNamespaces` is only updated by the NamespaceSelector.